File: docs/js/ed25519.js
// Ed25519 verify (RFC 8032) in pure JS — BigInt arithmetic + WebCrypto SHA-512.
// ESM. Vector-tested against the official RFC 8032 cases (tests/ed25519-vectors.mjs).

const P = (1n << 255n) - 19n;
const L = (1n << 252n) + 27742317777372353535851937790883648493n;
const D = mod(-121665n * modinv(121666n, P), P);
const SQRT_M1 = powmod(2n, (P - 1n) / 4n, P);
const BY = mod(4n * modinv(5n, P), P);
const BX = xrecover(BY);
const BASE = [BX, BY, 1n, mod((BX * BY), P)];

function mod(a, m) { return ((a % m) + m) % m; }
function powmod(b, e, m) { let r = 1n; b = mod(b, m); while (e > 0n) { if (e & 1n) r = mod(r * b, m); b = mod(b * b, m); e >>= 1n; } return r; }
function modinv(a, m) { return powmod(mod(a, m), m - 2n, m); }

function xrecover(y) {
  const xx = mod((y * y - 1n) * modinv(mod(D * y * y + 1n, P), P), P);
  let x = powmod(xx, (P + 3n) / 8n, P);
  if (mod(x * x - xx, P) !== 0n) x = mod(x * SQRT_M1, P);
  if ((x & 1n) !== 0n) x = P - x;
  return x;
}

function pointAdd(p1, p2) {
  const [x1, y1, z1, t1] = p1, [x2, y2, z2, t2] = p2;
  const a = mod((y1 - x1) * (y2 - x2), P);
  const b = mod((y1 + x1) * (y2 + x2), P);
  const c = mod(t1 * 2n * D * t2, P);
  const dd = mod(z1 * 2n * z2, P);
  const e = mod(b - a, P);
  const f = mod(dd - c, P);
  const g = mod(dd + c, P);
  const h = mod(b + a, P);
  return [mod(e * f, P), mod(g * h, P), mod(f * g, P), mod(e * h, P)];
}

function pointDbl(p1) {
  const [x1, y1, z1, t1] = p1;
  const a = mod(x1 * x1, P);
  const b = mod(y1 * y1, P);
  const c = mod(2n * z1 * z1, P);
  const d = mod(-a, P);
  const e = mod((x1 + y1) * (x1 + y1) - a - b, P);
  const g = mod(d + b, P);
  const f = mod(g - c, P);
  const h = mod(d - b, P);
  return [mod(e * f, P), mod(g * h, P), mod(f * g, P), mod(e * h, P)];
}

function scalarmult(point, scalar) {
  let q = [0n, 1n, 1n, 0n];
  for (let i = 0; i < 256; i++) {
    if ((scalar >> BigInt(i)) & 1n) q = pointAdd(q, point);
    point = pointDbl(point);
  }
  return q;
}

function encodePoint(p) {
  const [x, y, z] = p;
  const zi = modinv(z, P);
  const xx = mod(x * zi, P);
  const yy = mod(y * zi, P);
  const val = yy | ((xx & 1n) << 255n);
  const out = new Uint8Array(32);
  let v = val;
  for (let i = 0; i < 32; i++) { out[i] = Number(v & 255n); v >>= 8n; }
  return out;
}

function decodePoint(u8) {
  const raw = u8ToBig(u8);
  const y = raw & ((1n << 255n) - 1n);
  let x = xrecover(y);
  if ((x & 1n) !== ((raw >> 255n) & 1n)) x = P - x;
  return [mod(x, P), mod(y, P), 1n, mod(x * y, P)];
}

function u8ToBig(u8) {
  let x = 0n;
  for (let i = u8.length - 1; i >= 0; i--) x = (x << 8n) | BigInt(u8[i]);
  return x;
}

function concat(...arrays) {
  const total = arrays.reduce((n, a) => n + a.length, 0);
  const out = new Uint8Array(total);
  let off = 0;
  for (const a of arrays) { out.set(a, off); off += a.length; }
  return out;
}

function bytesEqual(a, b) {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
  return true;
}

async function sha512(u8) {
  return new Uint8Array(await globalThis.crypto.subtle.digest("SHA-512", u8));
}

export async function verify(pub, message, sig) {
  if (!(pub instanceof Uint8Array) || pub.length !== 32) return false;
  if (sig.length !== 64) return false;
  const renc = sig.slice(0, 32);
  const senc = sig.slice(32, 64);
  const s = u8ToBig(senc);
  if (s >= L) return false;
  let A, R;
  try {
    A = decodePoint(pub);
    R = decodePoint(renc);
  } catch (e) {
    return false;
  }
  const k = mod(u8ToBig(await sha512(concat(renc, pub, message))), L);
  const left = scalarmult(BASE, s);
  const right = pointAdd(R, scalarmult(A, k));
  return bytesEqual(encodePoint(left), encodePoint(right));
}

File: midsig/ed25519.py
"""Pure-Python Ed25519 (RFC 8032) — reference implementation.

Zero dependencies. Not constant-time — this is the portable reference used to
validate the wire format; production deployments should swap it for PyNaCl /
`cryptography` / OpenSSL, which expose the exact same 32-byte public keys and
64-byte signatures.
"""

import hashlib

P = 2**255 - 19
L = 2**252 + 27742317777372353535851937790883648493
D = (-121665 * pow(121666, -1, P)) % P
SQRT_M1 = pow(2, (P - 1) // 4, P)

# Base point B (extended homogeneous coordinates: X, Y, Z, T)
_BY = (4 * pow(5, -1, P)) % P
_BX = None


def _xrecover(y):
    xx = (y * y - 1) * pow(D * y * y + 1, -1, P) % P
    x = pow(xx, (P + 3) // 8, P)
    if (x * x - xx) % P != 0:
        x = (x * SQRT_M1) % P
    if x % 2 != 0:
        x = P - x
    return x


def _point_add(p1, p2):
    x1, y1, z1, t1 = p1
    x2, y2, z2, t2 = p2
    a = (y1 - x1) * (y2 - x2) % P
    b = (y1 + x1) * (y2 + x2) % P
    c = (t1 * 2 * D * t2) % P
    dd = (z1 * 2 * z2) % P
    e = (b - a) % P
    f = (dd - c) % P
    g = (dd + c) % P
    h = (b + a) % P
    return (e * f % P, g * h % P, f * g % P, e * h % P)


def _point_dbl(p1):
    x1, y1, z1, t1 = p1
    a = x1 * x1 % P
    b = y1 * y1 % P
    c = 2 * z1 * z1 % P
    d = -a % P
    e = ((x1 + y1) * (x1 + y1) - a - b) % P
    g = (d + b) % P
    f = (g - c) % P
    h = (d - b) % P
    return (e * f % P, g * h % P, f * g % P, e * h % P)


def _scalarmult(point, scalar):
    q = (0, 1, 1, 0)  # identity
    for i in range(256):
        if (scalar >> i) & 1:
            q = _point_add(q, point)
        point = _point_dbl(point)
    return q


def _basepoint():
    global _BX
    if _BX is None:
        _BX = _xrecover(_BY)
    return (_BX % P, _BY % P, 1, (_BX * _BY) % P)


def _encode_point(p):
    x, y, z, _ = p
    zi = pow(z, -1, P)
    x = x * zi % P
    y = y * zi % P
    return (y | ((x & 1) << 255)).to_bytes(32, "little")


def _decode_point(data):
    y = int.from_bytes(data, "little") & ((1 << 255) - 1)
    x = _xrecover(y)
    if (x & 1) != (int.from_bytes(data, "little") >> 255):
        x = P - x
    return (x % P, y % P, 1, (x * y) % P)


def _clamp_scalar(h):
    h = bytearray(h)
    h[0] &= 248
    h[31] &= 127
    h[31] |= 64
    return bytes(h)


def public_key(seed: bytes) -> bytes:
    """32-byte Ed25519 public key from a 32-byte seed."""
    if len(seed) != 32:
        raise ValueError("seed must be 32 bytes")
    h = hashlib.sha512(seed).digest()
    a = int.from_bytes(_clamp_scalar(h[:32]), "little")
    return _encode_point(_scalarmult(_basepoint(), a))


def sign(seed: bytes, message: bytes) -> bytes:
    """64-byte Ed25519 signature (R || S)."""
    if len(seed) != 32:
        raise ValueError("seed must be 32 bytes")
    h = hashlib.sha512(seed).digest()
    a = int.from_bytes(_clamp_scalar(h[:32]), "little")
    A = _scalarmult(_basepoint(), a)
    pub = _encode_point(A)
    r = int.from_bytes(hashlib.sha512(h[32:] + message).digest(), "little") % L
    R = _scalarmult(_basepoint(), r)
    renc = _encode_point(R)
    k = int.from_bytes(hashlib.sha512(renc + pub + message).digest(), "little") % L
    s = (r + k * a) % L
    return renc + s.to_bytes(32, "little")


def verify(pub: bytes, message: bytes, sig: bytes) -> bool:
    if len(pub) != 32 or len(sig) != 64:
        return False
    renc, senc = sig[:32], sig[32:]
    s = int.from_bytes(senc, "little")
    if s >= L:
        return False
    try:
        A = _decode_point(pub)
        R = _decode_point(renc)
    except (ValueError, ZeroDivisionError):
        return False
    k = int.from_bytes(hashlib.sha512(renc + pub + message).digest(), "little") % L
    left = _scalarmult(_basepoint(), s)
    right = _point_add(R, _scalarmult(A, k))
    return _encode_point(left) == _encode_point(right)
